cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/validator/CompileValidatorImpl.java

```diff
@@ -138,28 +133,73 @@ private boolean isForLoop(Step step, List<String> breakValuesList) {
-                for(Result result : refResults){
-                    String resultName = result.getName();
-                    if (!navigationListContainsKey(stepNavigationStrings, resultName)){
-                        errors.add(new IllegalArgumentException("Cannot compile flow: \'" + flow.getName() +
-                                "\' since for step: '" + step.getName() + "\', the result \'" + resultName+
-                                "\' of its dependency: \'"+ refId + "\' has no matching navigation"));
-                    }
-                }
+        String refId = step.getRefId();
+        if (reference == null) {
+            errors.add(new IllegalArgumentException(
+                    getErrorMessagePrefix(flow, step) + " the dependency '" + refId + "' is missing."
+            ));
+        } else {
+            if (!step.isOnFailureStep()) { // on_failure step cannot have navigation section
+                validateResultNamesAndNavigationSection(flow, step, refId, reference, errors);
             }
         }
         return errors;
     }
 
+    private void validateResultNamesAndNavigationSection(Flow flow, Step step, String refId, Executable reference, List<RuntimeException> errors) {
+        List<String> stepNavigationKeys = mapMapToKey(step.getNavigationStrings());
+        List<String> refResults = getReferenceResultNames(step, reference.getResults());
+
+        List<String> stepNavigationKeysWithoutMatchingResult = ListUtils.subtract(stepNavigationKeys, refResults);
+        List<String> refResultsWithoutMatchingNavigation = ListUtils.subtract(refResults, stepNavigationKeys);
+
+        if (CollectionUtils.isNotEmpty(refResultsWithoutMatchingNavigation)) {
+            errors.add(new IllegalArgumentException(
+                    getErrorMessagePrefix(flow, step) + " the results " + refResultsWithoutMatchingNavigation +
+                    " of its dependency '"+ refId + "' have no matching navigation."
+            ));
+        }
+        if (CollectionUtils.isNotEmpty(stepNavigationKeysWithoutMatchingResult)) {
+            errors.add(new IllegalArgumentException(
+                    getErrorMessagePrefix(flow, step) + " the navigation keys " +
+                    stepNavigationKeysWithoutMatchingResult + " have no matching results in its dependency '" +
+                    refId + "'."
+            ));
+        }
+    }
+
+    private String getErrorMessagePrefix(Flow flow, Step step) {
+        return "Cannot compile flow '" + flow.getName() +
+                "' since for step '" + step.getName() + "'";
+    }
+
+    private List<String> mapMapToKey(List<Map<String, String>> collection) {
+        List<String> result = new ArrayList<>();
+        for (Map<String, String> element : collection) {
+            result.add(element.keySet().iterator().next());
+        }
+        return result;
+    }
+
+    private List<String> mapResultsToNames(List<Result> results) {
```

Comment:
that's also perfectly valid - but this method is actually a clear lambda expression (in python it would be ~ `map(lambda x : x['name'], results))`, so I would like to stay with this naming (it will be easier to spot when we move to Java 8)